coreAPI/src/main/java/net/java/games/input/ControllerEnvironment.java

```diff
@@ -158,4 +164,46 @@ protected void fireControllerRemoved(Controller c) {
     public static ControllerEnvironment getDefaultEnvironment() {
         return defaultEnvironment;
     }
-}
\ No newline at end of file
+
+    public static ControllerEnvironment refreshDefaultEnvironment() {
+        DefaultControllerEnvironment newControllerEnvironment = new DefaultControllerEnvironment();
+        int oldControllerListLength = defaultEnvironment.getControllers().length;
+        int newControllerListLength = newControllerEnvironment.getControllers().length;
+
+        // If a controller is removed, the remaining controllers in the old environment should still be valid
+        // so the new environment can simply be discarded. Otherwise, the controllers in the old environment
+        // should be disposed before returning the new environment.
+        if (newControllerListLength < oldControllerListLength) {
+            for (int i = 0; i < newControllerListLength; i++) {
+                if (newControllerEnvironment.controllers.get(i) instanceof DisposableDevice) {
+                    try {
+                        ((DisposableDevice) newControllerEnvironment.controllers.get(i)).close();
+                    } catch (IOException e) {
+                        log("Device file descriptor is already closed: " + e.getMessage());
+                    }
+                }
+            }
+        } else {
+            for (int i = 0; i < newControllerListLength; i++) {
+                if (i < oldControllerListLength) {
+                    if (!((DefaultControllerEnvironment) defaultEnvironment).controllers.get(i).poll()) {
+                        if (((DefaultControllerEnvironment) defaultEnvironment).controllers.get(i) instanceof DisposableDevice) {
+                            try {
+                                ((DisposableDevice) ((DefaultControllerEnvironment) defaultEnvironment).controllers.get(i)).close();
+                            } catch (IOException e) {
+                                log("Device file descriptor is already closed: " + e.getMessage());
+                            }
+                        }
+                    }
+                }
+            }
+            // defaultEnvironment = newControllerEnvironment;   // Not sure why this was placed here???
+        }
```

Comment:
I found this a bit tough to follow with all of the nesting. Though I may have misunderstood it, would something like this work?

```java
final Set<Controller> newControllers = new HashSet<>(newControllers);
final Set<Controller> oldControllers = new HashSet<>(oldControllers);

final Set<Controller> addedControllers = newControllers.stream()
        .filter(newController -> !oldControllers.contains(newController))
        .collect(Collectors.toSet());

final Set<Controller> removedControllers = oldControllers.stream()
        .filter(oldController -> !newControllers.contains(oldController))
        .collect(Collectors.toSet());

// todo Add all addedControllers to the set of controllers in the old environment.

// todo Close all removedControllers.
```

And for the log message, maybe it should be set to `debug` as I'm unsure if it would be useful to a downstream dev or if it would be more noise in their logs

Reply:
Seems like that should work too, I wanted to stay true to the original PR I grabbed that from tho. I can test more on Monday perhaps. 😅